### PR TITLE
Cancel work order

### PR DESCRIPTION
**Pull request by pudiva**
Trello:
https://trello.com/c/LBgkvQFo/175-as-an-rcc-agent-i-need-to-be-able-to-cancel-the-work-order-if-information-changes-or-i-make-a-mistake

### CODE REVIEW

File: app/views/work_orders/cancellations/new.erb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            <%= @work_order.problem_description %>
+          </td>
+        </tr>
+      </table>
+
+      <!--
+      <h2 class="govuk-heading-m">
+        Reason to cancel
+      </h2>
+      -->
```
Comment (cdccollins) [inline app/views/work_orders/cancellations/new.erb]: why is this here?

File: app/views/work_orders/cancellations/new.erb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+      <%= govuk_form_with(
+        model: @cancellation,
+        url: work_order_cancellation_path(@work_order.reference),
+        local: true
+      ) do |form|
+      %>
+
+        <%= form.form_group :reason do %>
+          <%= form.label :reason, "Reason to cancel" %>
```
Comment (cdccollins) [inline app/views/work_orders/cancellations/new.erb]: I guess this is a design thing, but this isn't great English. Should be something like "Reason for cancelling"

File: spec/features/work_order_cancellation_spec.rb [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+  scenario 'Cancel a Work Order correctly', :js do
+    sign_in
+
+    # FIXME: s/hackneyrepairs/#{ ENV["..."] }/g
+
+    #
+    # stub minimum possible
+    #
+    stub_request(:get, "https://hackneyrepairs/v1/work_orders/00000666")
```
Comment (cdccollins) [inline spec/features/work_order_cancellation_spec.rb]: Might be nice to say what these are doing so if we come back to it we know what's going on rather than working it out all over again?

File: app/views/work_orders/cancellations/new.erb (hunk 2 above)
Comment (pudiva) [reply app/views/work_orders/cancellations/new.erb]: I'll tell Boris

File: spec/features/work_order_cancellation_spec.rb (hunk 3 above)
Comment (pudiva) [reply spec/features/work_order_cancellation_spec.rb]: what do you mean? like saying `# get work order 00000666` and so on?

Comment (cdccollins) [reply spec/features/work_order_cancellation_spec.rb]: yeah. I guess not super important but if stuff changes it's going to be a nightmare going back through those

Comment (pudiva) [reply spec/features/work_order_cancellation_spec.rb]: done! do you think it's enough?

File: app/views/work_orders/cancellations/new.erb (hunk 1 above)
Comment (pudiva) [reply app/views/work_orders/cancellations/new.erb]: removed :eyes:

File: spec/features/work_order_cancellation_spec.rb (hunk 3 above)
Comment (cdccollins) [reply spec/features/work_order_cancellation_spec.rb]: Looks good to me!

**Comment by pudiva** [comment]
@cdccollins can you please review those new commits? Sorry I know it's painful.

**Comment by cdccollins** [comment]
> @cdccollins can you please review those new commits? Sorry I know it's painful.

Yep, was just waiting for it all to be done :)

File: app/views/work_orders/cancellations/new.erb (hunk 2 above)
Comment (pudiva) [reply app/views/work_orders/cancellations/new.erb]: he agrees :sparkles:

**Comment by pudiva** [comment]
@cdccollins can you please review this? It should be ready now!

**Comment by cdccollins** [comment]
> @cdccollins can you please review this? It should be ready now!

I'm now getting this error when I try to raise a work order (so I can cancel it
Screenshot 2019-07-02 at 17 21 12
)

**Comment by cdccollins** [comment]
> @cdccollins can you please review this? It should be ready now!

I'm getting this error, have dockered and re-dockered 

Screenshot 2019-07-02 at 17 30 34

**Comment by pudiva** [comment]
> > @cdccollins can you please review this? It should be ready now!
> 
> I'm getting this error, have dockered and re-dockered
> 
> Screenshot 2019-07-02 at 17 30 34

This is because they changed the API response when your account is not associated in UH. Now it doesn't return an array of errors, but only an object.

Just override the email on application_controller for now. I'm fixing the error parsing.

**Comment by pudiva** [comment]
After some hours of debugging:

```
{"message": "Endpoint request timed out"}
```

but it works on my machine for some reason. Maybe because I'm on Hackney wired network?

**Comment by pudiva** [comment]
In the end it was just the API timeouting when the email is correct.

**Comment by pudiva** [comment]
Ok, this should be good now!

File: app/views/work_orders/cancellations/new.erb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+          <span class="govuk-hint">
+            You have <span id="counter" data-maximum-length="200">200</span> characters remaining.
+          </span>
+        <% end %>
+
+        <% if @work_order.is_dlo? && @work_order.servitor_reference.blank? %>
+          <%= render layout: "shared/govuk_warning_text" do %>
+            This works order does not have a Servitor reference. Please wait 30
+            minutes and then cancel this works order. If the work was due to take
+            place today. Please call the Planners before cancelling the Works
```
Comment (cdccollins) [inline app/views/work_orders/cancellations/new.erb]: `If the work was due to take place today.` isn't a full sentence - is it supposed to be attached to the next one? Also, the first `works order` isn't capitalised, but the second `Works Order` is

Comment (pudiva) [reply app/views/work_orders/cancellations/new.erb]: fixed!

Comment (pudiva) [reply app/views/work_orders/cancellations/new.erb]: :sparkles: :sparkles: :sparkles:

File: app/views/work_orders/cancellations/new.erb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          </span>
+        <% end %>
+
+        <% if @work_order.is_dlo? && @work_order.servitor_reference.blank? %>
+          <%= render layout: "shared/govuk_warning_text" do %>
+            This works order does not have a Servitor reference. Please wait 30
+            minutes and then cancel this Works Order. If the work was due to take
+            place today, please call the Planners before cancelling the Works
+            Order.
+          <% end %>
```
Comment (cdccollins) [inline app/views/work_orders/cancellations/new.erb:68]: 'works order' on line 64 doesn't match 'Works Order' on lines 66-67

Comment (pudiva) [reply app/views/work_orders/cancellations/new.erb:68]: should be good now!

**Review by cdccollins — APPROVED** [review]
